js/components/src/time-sync/time-sync.ts

```diff
@@ -0,0 +1,109 @@
+import { Platform } from "react-native";
+
+let timeOffset = 0;
+let hasWarned = false;
+let OriginalDate: DateConstructor = Date;
+
+const CLOCK_DRIFT_THRESHOLD_MS = 5000; // 5 seconds
+
+export function getTimeOffset(): number {
+  return timeOffset;
+}
+
+export function setTimeOffset(offset: number): void {
+  timeOffset = offset;
+}
+
+export function checkClockDrift(serverTime: string): {
+  hasDrift: boolean;
+  driftMs: number;
+  driftSeconds: number;
+} {
+  const serverDate = new Date(serverTime);
+  const clientDate = new Date();
+  const drift = Math.abs(serverDate.getTime() - clientDate.getTime());
+
+  if (drift > CLOCK_DRIFT_THRESHOLD_MS) {
+    const driftSeconds = Math.round(drift / 1000);
+    if (!hasWarned) {
+      hasWarned = true;
+      console.warn(
+        `clock drift detected: ${driftSeconds}s difference from server time. ` +
+          `this may cause issues with time-sensitive operations. ` +
+          `please sync your system clock.`,
+      );
+    }
+    return { hasDrift: true, driftMs: drift, driftSeconds };
+  } else {
+    return {
+      hasDrift: false,
+      driftMs: drift,
+      driftSeconds: Math.round(drift / 1000),
+    };
+  }
+}
+
+export function syncTimeWithServer(serverTime: string): void {
+  const serverDate = new Date(serverTime);
+  const clientDate = new Date();
+  const offset = serverDate.getTime() - clientDate.getTime();
+
+  setTimeOffset(offset);
+}
+
+export function getSyncedDate(): Date {
+  const now = new Date();
+  if (timeOffset !== 0) {
+    return new Date(now.getTime() + timeOffset);
+  }
+  return now;
+}
+
+export function getSystemDate(): Date {
+  return new OriginalDate();
+}
+
+export function getSystemTime(): number {
+  return OriginalDate.now();
+}
+
+export function initializeTimeSync(): void {
+  if (Platform.OS !== "web") {
+    return;
+  }
+
+  // store original Date
+  OriginalDate = Date;
```

Comment:
This makes me nervous because if initializeTimeSync gets called more than once we're gonna double-wrap it. Maybe put in a one-time static flag to check it's not called more than once?